Vulkan Multiview support for pixel shader debugging

Zorro666 opened the pull request:
## Description
- Implement Issue #3097 
  - Output `ViewIndex` into the pixel shader hits array
  - Select the pixel shader input which matches the `viewIndex` to the passed in `slice` parameter
    - uses the first colour attachment image view to convert `slice` to `viewIndex` 
  - Ignore the view parameter if set to ~0U (`rd.ReplayController.NoPreference`) 
  - Ignore the view parameter if any subpass has an empty multiviews

## Testing
- Tested using Sascha Williams `multiview` sample
- New Automated Test `VK_Multi_View`
  - colour attachment has four array slices and the imageView starts at slice array index of one
  - Vertex shader output Red for view 0, Green for view 1, fragment shader is pass thru and does not use viewIndex
  - Fragment shader output Red for view 0, Green for view 1, vertex shader does not use viewIndex
  - Geometry shader output Red for view 0, Green for view 1, fragment shader is pass thru and does not use viewIndex
  - Vertex shader output fixed colour to pass thru fragment, neither shader uses viewIndex 
  - Not implemented (because Mesh Shader does not currently support multiview): Mesh shader output Red for view 0, Green for view 1, fragment shader does not use viewIndex 
  - Python checks vertex and pixel shader debug output against replay rendering output
- Ran D3D11_, D3D12_, VK_, GL_ automated tests locally to check the python compiles with the extra required parameter for `DebugPixel` including slow tests and check Iter-Test and Repeat-Load

Zorro666 committed:
Vulkan DebugPixel support for multiview debugging

Output "ViewIndex" into the pixel shader hits array
Select the pixel shader input which matches the viewIndex to the passed in view
Ignore the view parameter if set to ~0U (rd.ReplayController.NoPreference)
Ignore the view parameter if any subpass has empty multiviews

Zorro666 committed:
Vulkan Multi-view test

Vertex shader output Red for view 0, Green for view 1, fragment shader does not use viewIndex.
Fragment shader output Red for view 0, Green for view 1, vertex shader does not use viewIndex.
Geometry shader output Red for view 0, Green for view 1, vertex and fragment shader do not use viewIndex.

Python checks vertex and pixel shader debug output against replay rendering output

baldurk commented on qrenderdoc/Windows/ShaderMessageViewer.cpp:
```diff
@@ -342,8 +342,8 @@ ShaderMessageViewer::ShaderMessageViewer(ICaptureContext &ctx, ShaderStageMask s
-          trace = r->DebugPixel(msg.location.pixel.x, msg.location.pixel.y,
-                                msg.location.pixel.sample, msg.location.pixel.primitive);
+          trace = r->DebugPixel(msg.location.pixel.x, msg.location.pixel.y, msg.location.pixel.sample,
+                                msg.location.pixel.primitive, msg.location.pixel.view, ~0U);
```
This is passing the view as a slice, but without performing any calculation to translate from one to the other.

baldurk commented on util/test/demos/vk/vk_multi_view.cpp:
```diff
@@ -0,0 +1,316 @@
+        0, VK_SUBPASS_EXTERNAL, VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT,
+        VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT,
+        VK_ACCESS_COLOR_ATTACHMENT_READ_BIT | VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT,
+        VK_ACCESS_MEMORY_READ_BIT, VK_DEPENDENCY_BY_REGION_BIT));
```
What's the reason for these subpass dependencies? Is there a VU that's not obvious which requires it?

baldurk commented on util/test/demos/vk/vk_multi_view.cpp:
```diff
@@ -0,0 +1,316 @@
+    rpMultiviewCreateInfo.correlationMaskCount = 1;
+    rpMultiviewCreateInfo.pCorrelationMasks = &correlationMask;
+
+    rpCreateInfo.pNext = &rpMultiviewCreateInfo;
```
The helpers usually define a function like

```
  RenderPassBeginInfo &next(const void *next)
  {
    this->pNext = next;
    return *this;
  }
```

for this, rather than casting & manually setting. You can add one to `vkh::RenderPassCreator`.

baldurk commented on util/test/demos/vk/vk_multi_view.cpp:
```diff
@@ -0,0 +1,316 @@
+
+      // Render multiview to its own framebuffer
+      const VkRenderPassBeginInfo *rpBeginInfo = vkh::RenderPassBeginInfo(
+          renderPass, framebuffer, mainWindow->scissor, {vkh::ClearValue(0.2f, 0.3f, 0.4f, 1.0f)});
```
One of the limitations of how I've defined the `vkh` helpers is they take things like `std::vector` parameters for arrays and directly reference them rather than storing them into local temporaries, which is convenient but means there is a hidden gotcha that they can't be used as temporary variables like this - the array of clear values has gone out of scope and now points to invalid memory. The `vkh::RenderPassBeginInfo` needs to be used in-line with the `vkCmdBeginRenderPass` as in other demos.

baldurk commented on renderdoc/driver/vulkan/vk_shaderdebug.cpp:
```diff
@@ -4296,6 +4371,91 @@ ShaderDebugTrace *VulkanReplay::DebugPixel(uint32_t eventId, uint32_t x, uint32_
+      useViewIndex = pipe.viewMask != 0;
+      if(!useViewIndex)
+        if(Vulkan_Debug_ShaderDebugLogging())
+          RDCLOG("Disabling useViewIndex because viewMask is zero");
```
To avoid ambiguity, please combine into a single depth `if` for this, or else use `{}` explicitly on all but the last nesting level.

baldurk commented on renderdoc/driver/vulkan/vk_shaderdebug.cpp:
```diff
@@ -4296,6 +4371,91 @@ ShaderDebugTrace *VulkanReplay::DebugPixel(uint32_t eventId, uint32_t x, uint32_
+  {
+    ResourceId rp = state.GetRenderPass();
+    ResourceId viewId;
+    // Need to convert the input slice into a viewIndex using the first colour attachment imageView
```
Unfortunately this is something I hadn't thought of earlier, but the problem here with this conversion of slice to view is that it's ambiguous - it's possible for different attachments to have different mappings, you can't always use attachment 0. You could have the colour attachment 0 where it only has N layers for N views and the mapping is 1:1, and colour attachment 1 is a view where it has N*2 layers and the mapping is non-identity. (Also as an aside, this would break debugging a pixel shader with only depth output as it expects a colour output).

In order to convert from view to slice here we'd need to know which image the user was thinking of when they specified the subresource. This is different to the existing sample index because all attachment images are required to have the same sample count and there's no conversion to a different identifier.

I think the options then are either to pass the image as well to DebugPixel so we can identify the view when needed, or pass the view separately from the slice and do the conversion in the UI with the context there. I'm struggling to think of another beneficial reason for the future to pass the image except to provide this context, so I'd lean towards just passing the view directly.

At this point also adding yet another parameter probably crosses the line into where we should have a config struct for debug pixel parameters - though I'd say keep EID & x/y as direct parameters since for convenience in scripts in many cases that's good enough.

baldurk reviewed (changes requested):
I think this looks good overall but there's one particular problem with the slice -> view conversion at shader debug time.

Zorro666 committed:
Vk Tests added vkh::RenderPassCreator::next()

Zorro666 committed:
Change DebugPixel() to take a DebugShaderInputs struct

DebugShaderInputs struct contains existing the parameters:

sample
primitive

and a new parameter

view

Default constructor for DebugShaderInputs sets the parameters to NoPreference (~0U).
